From: IrfanMohammad <irfan@graphiant.com>
Date: Thu, 27 Jul 2023 12:50:30 +0000
Subject: descriptive lock debugs and tests

In case of a lock failure, it is more useful to have more details of the
failure, such as which CIDs are holding the lock etc.

Add tests for reader lock failures
---
 src/common.c                  |  39 +++--
 src/common.h                  |   2 +
 src/shm_ext.c                 |  37 +++-
 src/shm_sub.c                 |   6 +
 src/subscr.c                  |   2 +
 tests/files/test-leafref.yang |   7 +
 tests/test_lock_log.c         | 382 ++++++++++++++++++++++++++++++++++++++++++
 7 files changed, 461 insertions(+), 14 deletions(-)
 create mode 100644 tests/test_lock_log.c

diff --git a/src/common.c b/src/common.c
index d2c3d9e..fe94bc5 100644
--- a/src/common.c
+++ b/src/common.c
@@ -64,10 +64,13 @@
 static int sr_is_priviliged(void);
 
 #define SR_RWLOCK_ID(rwlock_ptr) (~(1u<<31) & (rwlock_ptr)->id)
-#define _SR_LOG_RWLOCK(rwlock_ptr, mode, cid, func, str)                   \
-    SR_LOG_DBG("%s: %s rwlock %" PRIu32 " in mode %d by CID %" PRIu32, \
+#define _SR_LOG_RWLOCK(ll, rwlock_ptr, mode, cid, func, str)           \
+    sr_log(ll, "%s: %s rwlock %" PRIu32 " in mode %d by CID %" PRIu32, \
             func, str, SR_RWLOCK_ID(rwlock_ptr), mode, cid)
 
+#define SR_LOG_RWLOCK(rwlock_ptr, ...) sr_log_rwlock_enabled(rwlock_ptr) ? \
+    sr_log_rwlock_details(SR_LL_INF, rwlock_ptr, __VA_ARGS__) : (void)0
+
 static int
 sr_log_rwlock_enabled(sr_rwlock_t *rwlock)
 {
@@ -83,10 +86,9 @@ sr_log_rwlock_enabled(sr_rwlock_t *rwlock)
     return (sr_log_rwlock_is_enabled = (getenv("SR_LOG_RWLOCK") != NULL));
 }
 
-#define SR_LOG_RWLOCK(rwlock_ptr, ...) sr_log_rwlock_enabled(rwlock_ptr) ? sr_log_rwlock_details(rwlock_ptr, __VA_ARGS__) : (void)0
-
-static void
-sr_log_rwlock_details(sr_rwlock_t *rwlock, sr_lock_mode_t mode, sr_cid_t cid, const char *func_name, const char *msg)
+void
+sr_log_rwlock_details(sr_log_level_t log_level, sr_rwlock_t *rwlock, sr_lock_mode_t mode,
+        sr_cid_t cid, const char *func_name, const char *msg)
 {
     size_t i, cursor = 0;
     char buf[512] = "";
@@ -98,7 +100,7 @@ sr_log_rwlock_details(sr_rwlock_t *rwlock, sr_lock_mode_t mode, sr_cid_t cid, co
         }
     }
 
-    _SR_LOG_RWLOCK(rwlock, mode, cid, func_name, buf);
+    _SR_LOG_RWLOCK(log_level, rwlock, mode, cid, func_name, buf);
 }
 
 /**
@@ -2250,9 +2252,14 @@ sr_get_log_file(void)
     static FILE *logfile = NULL;
     int logfd = -1;
     struct stat fstats;
+    int ret;
 
     if (!logfile) {
-        asprintf(&logpath, "%s/rwlock_info", sr_get_repo_path());
+        ret = asprintf(&logpath, "%s/rwlock_info", sr_get_repo_path());
+        if (ret < 0) {
+            SR_LOG_WRN("%s: asprintf failed", __func__);
+            return NULL;
+        }
         logfd = sr_open(logpath, O_WRONLY | O_CREAT | O_APPEND, SR_SHM_PERM);
         if (logfd < 0) {
             SR_LOG_WRN("%s Unable to open logpath for writing errno %d", __func__, errno);
@@ -2282,6 +2289,7 @@ sr_rwlock_create_debug_info(uint32_t id, const char *desc)
     extern char * __progname;
     struct timespec ts = {0};
     clock_gettime(CLOCK_REALTIME, &ts);
+
     logfile = sr_get_log_file();
     if (!logfile) {
         return;
@@ -2340,7 +2348,7 @@ sr_rwlock_reader_add(sr_rwlock_t *rwlock, sr_cid_t cid, const char *func)
     if (i == SR_RWLOCK_READ_LIMIT) {
         sr_errinfo_new(&err_info, SR_ERR_TIME_OUT, "reader limit reached for lock %" PRIu32
                     " cid %" PRIu32, SR_RWLOCK_ID(rwlock), cid);
-        sr_log_rwlock_details(rwlock, SR_LOCK_READ, cid, func, "reader limit");
+        SR_RWLOCK_WRN(rwlock, SR_LOCK_READ, cid, func, "reader limit");
         goto cleanup;
     }
 
@@ -2353,7 +2361,7 @@ sr_rwlock_reader_add(sr_rwlock_t *rwlock, sr_cid_t cid, const char *func)
         if (rwlock->read_count[i] == UINT8_MAX) {
             sr_errinfo_new(&err_info, SR_ERR_INTERNAL, "Recursive reader limit reached for lock %" PRIu32
                     " cid %" PRIu32, SR_RWLOCK_ID(rwlock), cid);
-            sr_log_rwlock_details(rwlock, SR_LOCK_READ, cid, func, "reader recursive limit");
+            SR_RWLOCK_WRN(rwlock, SR_LOCK_READ, cid, func, "reader recursive limit");
             goto cleanup;
         }
         ++rwlock->read_count[i];
@@ -2428,6 +2436,7 @@ sr_rwlock_reader_del(sr_rwlock_t *rwlock, sr_cid_t cid)
     for (i = 0; SR_IS_READER_VALID(rwlock, i) && (rwlock->readers[i] != cid); ++i) {}
     if ((i == SR_RWLOCK_READ_LIMIT) || (rwlock->readers[i] != cid)) {
         /* CID not found */
+        SR_RWLOCK_WRN(rwlock, SR_LOCK_READ, cid, __func__, "CID not found");
         SR_ERRINFO_INT(&err_info);
         goto cleanup;
     }
@@ -2523,6 +2532,7 @@ sr_sub_rwlock(sr_rwlock_t *rwlock, struct timespec *timeout_abs, sr_lock_mode_t
         sr_rwlock_recover(rwlock, func, cb, cb_data);
         SR_CHECK_INT_RET(ret, err_info);
     } else if (ret) {
+        SR_RWLOCK_WRN(rwlock, mode, cid, func, "Locking");
         SR_ERRINFO_LOCK(&err_info, func, ret);
         return err_info;
     }
@@ -2675,7 +2685,7 @@ sr_sub_rwlock(sr_rwlock_t *rwlock, struct timespec *timeout_abs, sr_lock_mode_t
         }
     }
     if (err_info) {
-        sr_log_rwlock_details(rwlock, mode, cid, func, "Failed");
+        SR_RWLOCK_WRN(rwlock, mode, cid, func, "Failed");
     } else {
         SR_LOG_RWLOCK(rwlock, mode, cid, func, "Locked");
     }
@@ -2686,6 +2696,7 @@ error_cond_unlock:
         /* MUTEX UNLOCK */
         pthread_mutex_unlock(&rwlock->mutex);
     }
+    SR_RWLOCK_WRN(rwlock, mode, cid, func, "Condition");
 
     SR_ERRINFO_COND(&err_info, func, ret);
     return err_info;
@@ -2731,6 +2742,7 @@ sr_rwlock_mutex_lock(sr_rwlock_t *rwlock, uint32_t timeout_ms, const char *func,
         sr_rwlock_recover(rwlock, func, cb, cb_data);
         SR_CHECK_INT_RET(ret, err_info);
     } else if (ret) {
+        SR_RWLOCK_WRN(rwlock, SR_LOCK_NONE, 0, func, "mutex");
         SR_ERRINFO_LOCK(&err_info, func, ret);
         return err_info;
     }
@@ -2782,6 +2794,7 @@ sr_rwrelock(sr_rwlock_t *rwlock, uint32_t timeout_ms, sr_lock_mode_t mode, sr_ci
             }
         }
         if (ret) {
+            SR_RWLOCK_WRN(rwlock, mode, cid, __func__, "Condition relock");
             SR_ERRINFO_COND(&err_info, func, ret);
             goto cleanup_unlock;
         }
@@ -2849,6 +2862,7 @@ sr_rwrelock(sr_rwlock_t *rwlock, uint32_t timeout_ms, sr_lock_mode_t mode, sr_ci
             }
             rwlock->upgr = cid;
 
+            SR_RWLOCK_WRN(rwlock, mode, cid, func, "Condition relock");
             SR_ERRINFO_COND(&err_info, func, ret);
             goto cleanup_unlock;
         }
@@ -2954,6 +2968,7 @@ sr_rwunlock(sr_rwlock_t *rwlock, uint32_t timeout_ms, sr_lock_mode_t mode, sr_ci
             /* recover the lock, no CB needed since when we are unlocking, the state is expected to be valid and cleared */
             sr_rwlock_recover(rwlock, func, NULL, NULL);
         } else if (ret) {
+            SR_RWLOCK_WRN(rwlock, mode, cid, func, "UnLocking");
             SR_ERRINFO_LOCK(&err_info, func, ret);
             sr_errinfo_free(&err_info);
         }
@@ -3127,7 +3142,7 @@ sr_conn_oper_cache_add(sr_conn_ctx_t *conn, uint32_t sub_id, const char *module_
     cache->path = strdup(path);
     SR_CHECK_MEM_GOTO(!cache->path, err_info, cleanup);
     rc = asprintf(&lock_descr, "oper_poll_cache %s", path);
-    if (rc) {
+    if (rc < 0) {
         lock_descr = NULL;
     }
     if ((err_info = sr_rwlock_init(&cache->data_lock, 0, SR_RWLOCK_NEW_ID(conn), lock_descr))) {
diff --git a/src/common.h b/src/common.h
index d76af70..6cbd3d9 100644
--- a/src/common.h
+++ b/src/common.h
@@ -69,6 +69,8 @@ struct srplg_ntf_s;
 /* reserved lock ids to be used before main_shm is opened/created for process local locks */
 #define SR_RWLOCK_RESERVED_ID_MAX 100
 
+#define SR_RWLOCK_WRN(rwlock, mode, cid, func_name, msg) sr_log_rwlock_details(SR_LL_WRN, rwlock, mode, cid, func_name, msg)
+
 /** all ext SHM item sizes will be aligned to this number; also represents the allocation unit (B) */
 #define SR_SHM_MEM_ALIGN 8
 
diff --git a/src/shm_ext.c b/src/shm_ext.c
index 9fef134..a54bf16 100644
--- a/src/shm_ext.c
+++ b/src/shm_ext.c
@@ -41,6 +41,8 @@
 #include "shm_sub.h"
 #include "subscr.h"
 
+FILE * sr_get_log_file(void);
+
 static int
 sr_shmext_log_enabled(void)
 {
@@ -519,15 +521,30 @@ sr_shmext_change_sub_add(sr_conn_ctx_t *conn, sr_mod_t *shm_mod, sr_lock_mode_t
     off_t xpath_off;
     sr_mod_change_sub_t *shm_sub;
     uint32_t i;
+    FILE *logfile = sr_get_log_file();
+    extern char * __progname;
 
     /* kept for possible future use */
     assert(*has_lock == SR_LOCK_WRITE);
 
+    *xpath_id = ATOMIC_INC_RELAXED(SR_CONN_MAIN_SHM(conn)->new_xpath_id);
+
+    if (logfile) {
+        fprintf(logfile, "[%u][%u] [%s] change_sub CID %u module %s xpath_id %u xpath %s opts 0x%X priority %u evpipe %u\n",
+                getpid(), gettid(), __progname,
+                conn->cid, conn->mod_shm.addr + shm_mod->name, *xpath_id,
+                xpath ? xpath : "NULL", sub_opts, priority, evpipe_num);
+        fflush(logfile);
+    }
+
+    SR_LOG_DBG("change_sub CID %u module %s xpath_id %u xpath %s opts 0x%X priority %u evpipe %u",
+                conn->cid, conn->mod_shm.addr + shm_mod->name, *xpath_id,
+                xpath ? xpath : "NULL", sub_opts, priority, evpipe_num);
+
     /* EXT WRITE LOCK */
     if ((err_info = sr_shmext_conn_remap_lock(conn, SR_LOCK_WRITE, 1, __func__))) {
         goto cleanup;
     }
-    *xpath_id = ATOMIC_INC_RELAXED(SR_CONN_MAIN_SHM(conn)->new_xpath_id);
 
     if (sub_opts & SR_SUBSCR_UPDATE) {
         /* check that there is not already an update subscription with the same priority */
@@ -877,13 +894,29 @@ sr_shmext_oper_get_sub_add(sr_conn_ctx_t *conn, sr_mod_t *shm_mod, uint32_t sub_
     uint32_t i, j;
     int xpath_found = 0;
 
+    FILE *logfile = sr_get_log_file();
+    extern char * __progname;
+
     assert(path && sub_type);
 
+    *xpath_id = ATOMIC_INC_RELAXED(SR_CONN_MAIN_SHM(conn)->new_xpath_id);
+
+    if (logfile) {
+          fprintf(logfile, "[%u][%u] [%s] oper_get CID %u module %s xpath_id %u xpath %s opts 0x%X evpipe %u\n",
+                  getpid(), gettid(), __progname,
+                  conn->cid, conn->mod_shm.addr + shm_mod->name, *xpath_id,
+                  path, sub_opts, evpipe_num);
+          fflush(logfile);
+      }
+
+      SR_LOG_DBG("oper_get CID %u module %s xpath_id %u xpath %s opts 0x%X evpipe %u",
+                  conn->cid, conn->mod_shm.addr + shm_mod->name, *xpath_id,
+                  path, sub_opts, evpipe_num);
+
     /* EXT WRITE LOCK */
     if ((err_info = sr_shmext_conn_remap_lock(conn, SR_LOCK_WRITE, 1, __func__))) {
         return err_info;
     }
-    *xpath_id = ATOMIC_INC_RELAXED(SR_CONN_MAIN_SHM(conn)->new_xpath_id);
 
     /* check that this exact subscription does not exist yet while finding its position */
     new_len = sr_xpath_len_no_predicates(path);
diff --git a/src/shm_sub.c b/src/shm_sub.c
index 7002026..6f33917 100644
--- a/src/shm_sub.c
+++ b/src/shm_sub.c
@@ -46,6 +46,10 @@
 #include "sysrepo.h"
 #include "utils/nacm.h"
 
+extern void
+sr_log_rwlock_details(sr_log_level_t log_level, sr_rwlock_t *rwlock, sr_lock_mode_t mode,
+        sr_cid_t cid, const char *func_name, const char *msg);
+
 /**
  * @brief Generic structure for parallel notifications.
  */
@@ -400,6 +404,7 @@ sr_shmsub_notify_new_wrlock(sr_sub_shm_t *sub_shm, const char *shm_name, sr_sub_
                     "Waiting for subscription of \"%s\" failed, previous event \"%s\" ID %" PRIu32 " was not processed.",
                     shm_name, sr_ev2str(last_event), last_request_id);
         } else {
+            SR_RWLOCK_WRN(&sub_shm->lock, SR_LOCK_WRITE, cid, __func__, "Condition");
             /* other error */
             SR_ERRINFO_COND(&err_info, __func__, ret);
         }
@@ -496,6 +501,7 @@ _sr_shmsub_notify_wait_wr(sr_sub_shm_t *sub_shm, sr_sub_event_t event, uint32_t
             }
         } else {
             /* other error */
+            SR_RWLOCK_WRN(&sub_shm->lock, SR_LOCK_WRITE, cid, __func__, "Condition");
             SR_ERRINFO_COND(&err_info, __func__, ret);
         }
 
diff --git a/src/subscr.c b/src/subscr.c
index cd8062b..9f1b549 100644
--- a/src/subscr.c
+++ b/src/subscr.c
@@ -1423,8 +1423,10 @@ sr_rpc_sub_del(sr_subscription_ctx_t *subscr, uint32_t idx1, uint32_t idx2, sr_l
 cleanup_unlock:
     /* RPC SUB WRITE UNLOCK */
     if (is_ext) {
+        assert(shm_mod);
         sr_rwunlock(&shm_mod->rpc_ext_lock, 0, SR_LOCK_WRITE, subscr->conn->cid, __func__);
     } else {
+        assert(shm_rpc);
         sr_rwunlock(&shm_rpc->lock, 0, SR_LOCK_WRITE, subscr->conn->cid, __func__);
     }
 
diff --git a/tests/files/test-leafref.yang b/tests/files/test-leafref.yang
index f79db4a..4712a4c 100644
--- a/tests/files/test-leafref.yang
+++ b/tests/files/test-leafref.yang
@@ -31,6 +31,13 @@ module test-leafref {
         }
     }
 
+    leaf format-str-test {
+        type uint8;
+        must ". <= 1" {
+            error-message "This is a msg to test for format string failures %p %s %s %s %s %x %x %x %x % %";
+        }
+    }
+
     container car-state {
         config false;
         leaf a1 {
diff --git a/tests/test_lock_log.c b/tests/test_lock_log.c
new file mode 100644
index 0000000..4e75154
--- /dev/null
+++ b/tests/test_lock_log.c
@@ -0,0 +1,382 @@
+/**
+ * @file test_lock_log.c
+ * @author Irfan <irfan@graphiant.com>
+ * @brief test for sr_rwlock_t logging
+ *
+ * @copyright
+ * Copyright 2022 Graphiant
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#define _GNU_SOURCE
+
+#include <string.h>
+#include <unistd.h>
+#include <setjmp.h>
+#include <stdarg.h>
+#include <stdlib.h>
+#include <sys/types.h>
+#include <sys/wait.h>
+
+#include <cmocka.h>
+#include <libyang/libyang.h>
+
+#include "sysrepo.h"
+#include "common.h"
+#include "tests/tcommon.h"
+
+struct state {
+    sr_conn_ctx_t *conn;
+    sr_session_ctx_t *sess;
+    int conn_per_thread;
+    pthread_barrier_t barrier;
+};
+
+static int
+setup_f(void **state)
+{
+    struct state *st;
+    const char *schema_paths[] = {
+        TESTS_SRC_DIR "/files/test.yang",
+        TESTS_SRC_DIR "/files/test-leafref.yang",
+        NULL
+    };
+
+    st = calloc(1, sizeof *st);
+    *state = st;
+
+    if (sr_connect(0, &(st->conn)) != SR_ERR_OK) {
+        return 1;
+    }
+
+    if (sr_install_modules(st->conn, schema_paths, TESTS_SRC_DIR "/files", NULL) != SR_ERR_OK) {
+        return 1;
+    }
+    return 0;
+}
+static int
+teardown_f(void **state)
+{
+    struct state *st = (struct state *)*state;
+    const char *module_names[] = {
+        "test",
+        "test-leafref",
+        NULL
+    };
+
+    if (sr_remove_modules(st->conn, module_names, 0) != SR_ERR_OK) {
+        return 1;
+    }
+
+    sr_disconnect(st->conn);
+    free(st);
+    return 0;
+}
+
+typedef enum {
+    TEST_CB_SLEEP = 0x1,
+    TEST_CB_CRASH = 0x2,
+} test_cb_action_t;
+
+ATOMIC_T test_cb_action;
+ATOMIC_T test_cb_sleep_ms;
+
+static void
+test_cb_act(test_cb_action_t action, uint32_t delay)
+{
+    if (action & TEST_CB_SLEEP) {
+        TLOG_INF("%s sleeping %u ms", __func__, delay);
+        usleep(delay * 1000u);
+    }
+    if (action & TEST_CB_CRASH) {
+        TLOG_INF("%s Dying", __func__);
+        exit(0);
+    }
+}
+
+static int
+module_action_cb(sr_session_ctx_t *session, uint32_t sub_id, const char *module_name,
+        const char *xpath, sr_event_t event, uint32_t request_id, void *private_data)
+{
+    (void)module_name;
+    (void)sub_id;
+    (void)xpath;
+    (void)event;
+    (void)request_id;
+    (void)private_data;
+    (void)session;
+    test_cb_act(ATOMIC_LOAD_RELAXED(test_cb_action), ATOMIC_LOAD_RELAXED(test_cb_sleep_ms));
+    return 0;
+}
+
+static int
+test_oper_action_cb(sr_session_ctx_t *session, uint32_t sub_id, const char *module_name, const char *xpath,
+        const char *request_xpath, uint32_t request_id, struct lyd_node **parent, void *private_data)
+{
+    (void)session;
+    (void)sub_id;
+    (void)module_name;
+    (void)xpath;
+    (void)request_xpath;
+    (void)request_id;
+    (void)parent;
+    (void)private_data;
+    test_cb_act(ATOMIC_LOAD_RELAXED(test_cb_action), ATOMIC_LOAD_RELAXED(test_cb_sleep_ms));
+
+    return SR_ERR_OK;
+}
+
+static ATOMIC_T stop_threads;
+
+static void *
+test_lock_log_reader_limit_helper(void *arg)
+{
+    struct state *st = (struct state *)arg;
+    sr_conn_ctx_t *conn = NULL;
+    sr_session_ctx_t *sess = NULL;
+    sr_data_t *data = NULL;
+    int ret = 0;
+
+    if (st->conn_per_thread) {
+        ret = sr_connect(0, &conn);
+        assert_int_equal(ret, SR_ERR_OK);
+    } else {
+        conn = st->conn;
+    }
+
+    ret = sr_session_start(conn, SR_DS_OPERATIONAL, &sess);
+    assert_int_equal(ret, SR_ERR_OK);
+
+    /* weak allow all threads to spawn */
+    usleep(200000);
+    while (ret == SR_ERR_OK && !ATOMIC_LOAD_RELAXED(stop_threads)) {
+        data = NULL;
+        ret = sr_get_data(sess, "/test:*", 0, 0, 0, &data);
+        sched_yield();
+        sr_release_data(data);
+    }
+    TLOG_INF("%s Asking all threads to stop", __func__);
+    ATOMIC_STORE_RELAXED(stop_threads, 1);
+    sleep(1);
+    ret = sr_session_stop(sess);
+    assert_int_equal(ret, SR_ERR_OK);
+    if (st->conn_per_thread) {
+        sr_disconnect(conn);
+    }
+    return NULL;
+}
+
+static void
+test_lock_log_reader_limit(void **arg)
+{
+    struct state *st = (struct state *)*arg;
+    sr_subscription_ctx_t *subscr = NULL;
+    size_t i, j;
+    int ret;
+    sr_data_t *data = NULL;
+    const size_t NUM_THREADS = 2 + SR_RWLOCK_READ_LIMIT;
+    pthread_t tid[NUM_THREADS];
+
+    ATOMIC_STORE_RELAXED(stop_threads, 0);
+    sr_session_start(st->conn, SR_DS_OPERATIONAL, &st->sess);
+
+    ATOMIC_STORE_RELAXED(test_cb_action, TEST_CB_SLEEP);
+    ATOMIC_STORE_RELAXED(test_cb_sleep_ms, 10);
+
+    /* Create a subscription that will crash */
+    ret = sr_oper_get_subscribe(st->sess, "test", "/test:*", test_oper_action_cb, NULL, 0, &subscr);
+    assert_int_equal(ret, SR_ERR_OK);
+
+    for (j = 1; j < 2; j++) {
+        st->conn_per_thread = j;
+        /* Spawn more threads that will read than concurrent readers */
+        for (i = 0; i < NUM_THREADS; i++)
+        {
+            pthread_create(&tid[i], NULL, test_lock_log_reader_limit_helper, st);
+        }
+        sleep(1 + j*1);
+        ATOMIC_STORE_RELAXED(stop_threads, 1);
+        for (i = 0; i < NUM_THREADS; i++)
+        {
+            pthread_join(tid[i], NULL);
+        }
+
+        TLOG_INF("%s All threads finished", __func__);
+        /* Verify that after reader limit is reached, get succeeds eventually */
+        ret = sr_get_data(st->sess, "/test:*", 0, 0, 0, &data);
+        assert_int_equal(ret, SR_ERR_OK);
+        sr_release_data(data);
+    }
+
+    ret = sr_unsubscribe(subscr);
+    assert_int_equal(ret, SR_ERR_OK);
+    ret = sr_session_stop(st->sess);
+    assert_int_equal(ret, SR_ERR_OK);
+    st->sess = NULL;
+}
+
+static void
+test_lock_cb_timeout(void **arg)
+{
+    struct state *st = (struct state *)*arg;
+    sr_subscription_ctx_t *subscr = NULL;
+    int ret;
+
+    ret = sr_session_start(st->conn, SR_DS_RUNNING, &st->sess);
+    assert_int_equal(ret, 0);
+
+    ATOMIC_STORE_RELAXED(test_cb_action, TEST_CB_SLEEP);
+    ATOMIC_STORE_RELAXED(test_cb_sleep_ms, 100 + SR_SHMEXT_SUB_LOCK_TIMEOUT);
+
+    /* Start some subscriptions to test.yang */
+    ret = sr_module_change_subscribe(st->sess, "test", NULL, module_action_cb, NULL,
+            0, 0, &subscr);
+    assert_int_equal(ret, 0);
+
+    ret = sr_set_item_str(st->sess, "/test:l1[k='some-key']/v", "25", NULL, 0);
+    assert_int_equal(ret, SR_ERR_OK);
+
+    TLOG_INF("starting apply_changes");
+    ret = sr_apply_changes(st->sess, 100);
+
+    TLOG_INF("try unsubscribe, times out");
+    ret = sr_unsubscribe(subscr);
+    assert_int_equal(ret, SR_ERR_TIME_OUT);
+
+    TLOG_INF("starting unsubscribe again");
+    ret = sr_unsubscribe(subscr);
+    assert_int_equal(ret, SR_ERR_OK);
+    TLOG_INF("Succeeded unsubscribe on retry");
+
+    ret = sr_session_stop(st->sess);
+    assert_int_equal(ret, SR_ERR_OK);
+    st->sess = NULL;
+}
+
+static void
+test_err_log(void **arg)
+{
+    struct state *st = (struct state *)*arg;
+    int ret;
+
+    ret = sr_session_start(st->conn, SR_DS_RUNNING, &st->sess);
+    assert_int_equal(ret, 0);
+
+    /* This forces us to print a message that contains a % which can trigger a format string vulnerability */
+    ret = sr_set_item_str(st->sess, "/test-leafref:format-str-test", "120", NULL, 0);
+    assert_int_equal(ret, SR_ERR_OK);
+
+    ret = sr_apply_changes(st->sess, 100);
+    assert_int_equal(ret, SR_ERR_VALIDATION_FAILED);
+
+    ret = sr_session_stop(st->sess);
+    assert_int_equal(ret, SR_ERR_OK);
+    st->sess = NULL;
+}
+
+static void
+test_lock_dead_reader_helper(void)
+{
+    sr_conn_ctx_t *conn = NULL;
+    sr_session_ctx_t *sess = NULL;
+    sr_data_t *data = NULL;
+    sr_subscription_ctx_t *subscr = NULL;
+    char xpath[128] = "";
+
+    int ret;
+
+    ret = sr_connect(0, &conn);
+    assert_int_equal(ret, SR_ERR_OK);
+
+    ret = sr_session_start(conn, SR_DS_OPERATIONAL, &sess);
+    assert_int_equal(ret, 0);
+
+    snprintf(xpath, sizeof xpath, "/test-leafref:car-state/enabled-params[.='key%d']", getpid());
+
+    ATOMIC_STORE_RELAXED(test_cb_action, TEST_CB_SLEEP | TEST_CB_CRASH);
+    ATOMIC_STORE_RELAXED(test_cb_sleep_ms, 1000);
+
+    /* Create a subscription that will crash */
+    ret = sr_oper_get_subscribe(sess, "test-leafref", xpath, test_oper_action_cb, NULL, 0, &subscr);
+    assert_int_equal(ret, SR_ERR_OK);
+
+    /* weak wait until every process gets here */
+    sleep(1);
+    TLOG_INF("Getting data");
+    /* Now read our own data so we crash */
+    ret = sr_get_data(sess, xpath, 0, 0, 0, &data);
+
+    /* if we didn't crash in the callback,
+     * because a lock blocked us from getting there, we will crash now */
+    assert_int_equal(ret, -1);
+
+    sr_release_data(data);
+    data = NULL;
+
+    sr_disconnect(conn);
+}
+
+static void
+test_lock_dead_reader_limit(void **arg)
+{
+    struct state *st = (struct state *)*arg;
+    int i, ret;
+    const int NUM_PIDS = 2 * SR_RWLOCK_READ_LIMIT;
+    pid_t pid;
+    sr_data_t *data = NULL;
+
+    for (i = 0; i < NUM_PIDS; i++) {
+        switch (pid = fork()) {
+            case 0:
+                /* child */
+                test_lock_dead_reader_helper();
+                /* should never come here */
+                TLOG_WRN("%s: never come here", __func__);
+                abort();
+            default:
+                assert_int_not_equal(pid, -1);
+                break;
+        }
+    }
+    for (i = 0; i < NUM_PIDS; i++) {
+        wait(&ret);
+    }
+    /* Now that all the processes which were supposed to die, have died, ensure that we actually can read again */
+
+    TLOG_INF("Check that we can still read data");
+    ret = sr_session_start(st->conn, SR_DS_OPERATIONAL, &st->sess);
+    assert_int_equal(ret, SR_ERR_OK);
+    ret = sr_get_data(st->sess, "/*", 0, 0, 0, &data);
+    assert_int_equal(ret, SR_ERR_OK);
+
+    sr_release_data(data);
+
+    ret = sr_session_stop(st->sess);
+    assert_int_equal(ret, SR_ERR_OK);
+    st->sess = NULL;
+}
+
+int
+main(void)
+{
+    const struct CMUnitTest tests[] = {
+        cmocka_unit_test(test_lock_log_reader_limit),
+        cmocka_unit_test(test_err_log),
+        cmocka_unit_test(test_lock_cb_timeout),
+        cmocka_unit_test(test_lock_dead_reader_limit),
+    };
+    setenv("CMOCKA_TEST_ABORT", "1", 1);
+    setenv("SR_TEST_LOG_DEBUG", "1", 1);
+    test_log_init();
+    return cmocka_run_group_tests(tests, setup_f, teardown_f);
+}
